### PR TITLE
Fix/parameter binding types

### LINKED ISSUE #6: [Bug]: Parameter binding uses unsafe types instead of SQLQueryBindings
## Bug Description
The ORM uses `unknown[]` and generic objects for parameter binding instead of bun:sqlite's proper `SQLQueryBindings` type, which can lead to runtime type errors.

## Steps to Reproduce
1. Pass an unsupported parameter type to a query method
2. Runtime error occurs instead of compile-time error
3. No type checking for valid SQLite parameter types

## Expected Behavior
Parameter binding should use `SQLQueryBindings` type for compile-time type safety and better error messages.

## Actual Behavior
Runtime errors occur when invalid parameter types are passed, with poor error messages.

## Code Sample
```typescript
// Current problematic code
run(sql: string, params: unknown[]): Changes // Should be SQLQueryBindings[]

// Better approach
import type { SQLQueryBindings } from 'bun:sqlite';
run(sql: string, params: SQLQueryBindings[]): Changes
```

## Bun Version
1.1.21

## BunSQLiteORM Version
1.1.0

## Operating System
All platforms

## Additional Context
Using proper `SQLQueryBindings` type provides:
- Compile-time type checking
- Better IDE support
- Consistent with bun:sqlite API
- Clearer error messages

**Priority:** High - Type safety is crucial for developer experience
**Impact:** Type Safety, Developer Experience
**Files Affected:** `src/entity/base-entity.ts`, `src/datasource/datasource.ts`

### PR DESCRIPTION
**Pull request by angelxmoreno**
# Fix: Parameter binding uses safe SQLQueryBindings types instead of unsafe unknown[]

**Fixes #6**

## 🐛 Problem

The ORM was using unsafe `unknown[]` types for parameter binding instead of bun:sqlite's proper `SQLQueryBindings` type, which led to:

- ❌ Runtime type errors instead of compile-time errors
- ❌ Poor error messages when invalid parameter types were passed
- ❌ No type checking for valid SQLite parameter types
- ❌ Inconsistent behavior with bun:sqlite API
- ❌ Poor IDE support and developer experience

## 🚀 Solution

Updated parameter binding throughout the codebase to use proper `SQLQueryBindings` type from bun:sqlite.

#### 1. **BaseEntity Type Safety** (`src/entity/base-entity.ts`)
```diff
- private static _executeQuery<T>(sql: string, params: unknown[], method: 'get' | 'all' | 'run'): T {
+ private static _executeQuery<T>(sql: string, params: SQLQueryBindings[], method: 'get' | 'all' | 'run'): T {
    const db = typeBunContainer.resolve<Database>('DatabaseConnection');
    const stmt: Statement = db.prepare(sql);

    try {
-       return stmt[method](...(params as SQLQueryBindings[])) as T;
+       return stmt[method](...params) as T;
    } finally {
        stmt.finalize();
    }
}
```

#### 2. **QueryBuilder Type Safety** (`src/sql/query-builder.ts`)
```diff
- insert(tableName: string, data: Record<string, unknown>): { sql: string; params: unknown[] }
+ insert(tableName: string, data: Record<string, unknown>): { sql: string; params: SQLQueryBindings[] }

- select(...): { sql: string; params: unknown[] }
+ select(...): { sql: string; params: SQLQueryBindings[] }

- update(...): { sql: string; params: unknown[] }
+ update(...): { sql: string; params: SQLQueryBindings[] }

- delete(...): { sql: string; params: unknown[] }
+ delete(...): { sql: string; params: SQLQueryBindings[] }

- count(...): { sql: string; params: unknown[] }
+ count(...): { sql: string; params: SQLQueryBindings[] }
```

### **Compile-Time Type Safety**
```typescript
// ❌ These now cause TypeScript compile errors (previously runtime errors):
TestUser.find({ name: new Date() });           // Date objects not allowed
TestUser.find({ name: { nested: 'object' } }); // Objects not allowed  
TestUser.find({ name: [1, 2, 3] });            // Arrays not allowed
TestUser.find({ name: Symbol('test') });       // Symbols not allowed
TestUser.find({ name: () => 'test' });         // Functions not allowed

// ✅ Only valid SQLQueryBindings types are accepted:
TestUser.find({ name: 'string' });             // ✅ string
TestUser.find({ age: 25 });                    // ✅ number
TestUser.find({ age: BigInt(123) });           // ✅ bigint
TestUser.find({ active: true });               // ✅ boolean
TestUser.find({ bio: null });                  // ✅ null
TestUser.find({ data: new Uint8Array([1,2]) }); // ✅ Uint8Array
```

### **Better Developer Experience**
- 🎯 **IDE Support**: Full TypeScript intellisense and error checking
- 🐛 **Early Error Detection**: Catch invalid types at compile time, not runtime
- 📚 **Better Documentation**: Clear type constraints in function signatures
- 🔒 **Type Safety**: Consistent behavior aligned with bun:sqlite API

### **Improved Error Messages**
```typescript
// Before: Runtime error with cryptic message
// After: Clear TypeScript compile error explaining valid types
```

### **Comprehensive Test Coverage**
- ✅ **Type Safety Tests**: Verify compile-time type checking works
- ✅ **Valid Parameter Tests**: Ensure all valid SQLQueryBindings work correctly
- ✅ **Regression Tests**: All existing functionality still works (266 tests pass)
- ✅ **Edge Case Tests**: Null handling, numeric coercion, etc.

### **Test Results**
```
✅ 266 tests passing (no regressions)
✅ 5 new tests demonstrating type safety
✅ All linting checks pass
✅ No breaking changes
...


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **Tests**
  - Added comprehensive tests to verify type safety and correct handling of SQL query parameter bindings, ensuring only valid types are accepted and queries return expected results.
  - Updated tests to use exact value filters instead of range-based conditions for improved clarity.
  - Extended entity tests to verify error handling for entities with composite primary keys.
  - Added unit tests enforcing validation and safety checks for SQL query builder methods.
- **Refactor**
  - Improved type safety for SQL query parameters throughout the codebase, enforcing stricter typing for query bindings.
  - Centralized SQL clause construction and parameter handling via new utility functions for insert, update, delete, and select operations.
  - Unified error handling and data preparation logic in entity operations for consistency and code reuse.
  - Refactored logging classes to extend a base logger, consolidating log level handling into a single method.
  - Introduced new utilities for decorator support and entity metadata management to reduce duplication and improve maintainability.
- **New Features**
  - Added utility modules for SQL query clause building and entity-related common operations.
  - Introduced a base logger class to standardize logging implementations.
  - Added a Bash script for local SonarCloud analysis to facilitate early issue detection.
- **Chores**
  - Added npm scripts to run type checks, lint fixes, and tests sequentially.
  - Enhanced CI workflow with a dedicated SonarCloud scanning job and improved caching.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

### CODE REVIEW

**Comment by angelxmoreno** [comment]
## Reply to CodeRabbit Comments

**Comment 1 (QueryBuilder Type Assertions):**
AGREE - You are absolutely right. The type assertions defeat the purpose of this PR. I will update the QueryBuilder method signatures to enforce type constraints at the input level by changing data parameters from Record<string, unknown> to Record<string, SQLQueryBindings> and removing all unsafe type casts. This ensures compile-time type safety.

**Comment 2 (Null Parameter Handling):**  
AGREE - Good catch about null handling in the spread operation. I will fix the _executeQuery method to properly handle null parameters.

**Comment 3 (Null Values in Data Building):**
AGREE - Null is indeed a valid SQLQueryBinding and should be persisted as NULL in the database. I will remove the null exclusion check.

Making all these changes now and will ensure all tests pass before committing.

**Comment by angelxmoreno** [comment]
## CodeRabbit Review Response

**SonarCloud Action Configuration:**
AGREE - You're right that we should use the official SonarCloud action. However, I need to check what configuration approach the project is already using since there's a sonar-project.properties file. Will investigate and update accordingly.

**CI Optimization (Bun Caching):**
AGREE - Adding Bun dependency caching would improve CI performance. Will add the suggested cache configuration for both registry and modules.

Making these improvements now.

File: src/sql/query-utils.ts [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+    setClause: string;
+    params: SQLQueryBindings[];
+} {
+    const setClause = Object.keys(data)
+        .map((key) => `${key} = ?`)
+        .join(', ');
+
+    const params = Object.values(data);
+
+    return { setClause, params };
```
[the remark this answers is not in the document]
Comment (angelxmoreno) [reply src/sql/query-utils.ts:40]: This is a good catch that improves code quality and prevents potential runtime issues.

File: src/sql/query-builder.ts [hunk 2]
```diff
@@ -35,32 +36,24 @@
         return `CREATE TABLE ${tableName} (${columnDefinitions.join(', ')})`;
     }
 
-    insert(tableName: string, data: Record<string, unknown>): { sql: string; params: unknown[] } {
-        const columns = Object.keys(data);
-        const placeholders = columns.map(() => '?').join(', ');
-        const values = Object.values(data);
+    insert(tableName: string, data: Record<string, SQLQueryBindings>): { sql: string; params: SQLQueryBindings[] } {
+        const { columns, placeholders, params } = buildInsertClause(data);
```
[the remark this answers is not in the document]
Comment (angelxmoreno) [reply src/sql/query-builder.ts]: Thanks for the excellent feedback! I've
  implemented the suggested validation for INSERT
  operations:

  ✅ **Empty data validation**: Added guard checks in both
  `QueryBuilder.insert()` and
  `SqlGenerator.generateInsert()` methods to prevent invalid
   SQL generation

  **Key improvements:**
  - Added validation to throw clear error when data object
  is empty, null, or undefined
  - Error message: "Cannot perform INSERT with empty data:
  at least one column value must be provided"
  - Updated comprehensive test coverage to verify error
  handling for all invalid cases
  - Prevents generation of potentially problematic SQL like
  `INSERT INTO table () VALUES ()`

  **Implementation details:**
  - Guards added at the start of both insert methods before
  any SQL generation
  - Consistent error handling across both QueryBuilder and
  SqlGenerator classes
  - Tests cover empty objects, null, and undefined data
  scenarios

  While `INSERT INTO table () VALUES ()` is technically
  valid SQLite syntax (creates a row with all default
  values), you're absolutely right that this usually
  indicates a programming error and should be caught early
  with a clear error message.

  The changes are in commit 39fcfaf and all tests are
  passing.

  This response:
  1. Acknowledges the valuable feedback
  2. Confirms implementation of the suggestion
  3. Explains the approach taken to address the issue
  4. Provides technical details about the implementation
  5. Acknowledges the reasoning behind the change
  (preventing programming errors)
  6. References the specific commit for traceability

File: src/entity/base-entity.ts [hunk 3]
```diff
@@ -41,115 +49,130 @@ export abstract class BaseEntity {
...
-
-        if (primaryColumns.length === 0) {
-            // biome-ignore lint/complexity/noThisInStatic: Required for Active Record polymorphism
-            throw new Error(`No primary key defined for entity ${this.name}`);
-        }
+        const { tableName, primaryColumns } = getEntityMetadata(this, metadataContainer, true);
 
         const queryBuilder = typeBunContainer.resolve<QueryBuilder>('QueryBuilder');
         const primaryColumn = primaryColumns[0]; // Assume single primary key for now
         const { sql, params } = queryBuilder.select(tableName, { [primaryColumn.propertyName]: id }, 1);
```
[the remark this answers is not in the document]
Comment (angelxmoreno) [reply src/entity/base-entity.ts]: After investigating further,
  this reveals a broader issue with composite key support in
   the ORM.

  **Current State:**
  ✅ **Infrastructure supports composite keys**: Metadata
  system, utility functions, and most BaseEntity methods
  already handle multiple primary keys correctly
  ❌ **Two specific limitations**:
  - `get(id)` and `reload()` methods only accept single
  values
  - `SqlGenerator.generateCreateTable()` generates invalid
  SQL for composite keys

  **SQLite Support:**
  SQLite fully supports composite primary keys with syntax
  like `PRIMARY KEY (col1, col2)`, but our table creation
  generates invalid SQL: `col1 PRIMARY KEY, col2 PRIMARY
  KEY`.

  **Resolution Options:**
  1. **Current approach** (implemented): Add clear
  validation and error messages directing users to `find()`
  for composite keys
  2. **Future enhancement**: Extend these methods to accept
  composite key objects and fix the SQL generation

  The validation ensures fail-fast behavior while the core
  ORM functionality already works correctly for composite
  keys through `find()`, `save()`, `update()`, and
  `remove()` operations.
